test/versioned/pino/pino.tap.js

```diff
@@ -0,0 +1,320 @@
+/*
+ * Copyright 2021 New Relic Corporation. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+const tap = require('tap')
+const { sink, once } = require('pino/test/helper')
+const { truncate } = require('../../../lib/util/application-logging')
+const helper = require('../../lib/agent_helper')
+
+tap.Test.prototype.addAssert(
+  'validateAnnotations',
+  2,
+  function assertCoreAnnotations({ line: logLine, message, level, config }) {
+    this.equal(
+      logLine['entity.name'],
+      config.applications()[0],
+      'should have entity name that matches app'
+    )
+    this.equal(logLine['entity.guid'], 'pino-guid', 'should have set entitye guid')
+    this.equal(logLine['entity.type'], 'SERVICE', 'should have entity type of SERVICE')
+    this.equal(logLine.hostname, config.getHostnameSafe(), 'should have proper hostname')
+    this.match(logLine.timestamp, /[0-9]{10}/, 'should have proper unix timestamp')
+    this.notOk(logLine.message.includes('NR-LINKING'), 'should not contain NR-LINKING metadata')
+    if (message) {
+      this.equal(logLine.message, message, 'message should be the same as log')
+    }
+
+    if (level) {
+      this.equal(logLine.level, level, 'level should be string value not number')
+    }
+  }
+)
+
+// NOTE: pino adds hostname to log lines which is why we don't check it here
+const contextKeys = ['entity.name', 'entity.type', 'entity.guid', 'trace.id', 'span.id']
+
+tap.test('Pino instrumentation', (t) => {
+  t.autoend()
+  let logger
+  let stream
+  let pino
+  let agent
+
+  function setupAgent(config) {
+    agent = helper.instrumentMockedAgent(config)
+    agent.config.entity_guid = 'pino-guid'
+    pino = require('pino')
+    stream = sink()
+    logger = pino({ level: 'debug' }, stream)
+    return agent.config
+  }
+
+  t.afterEach(() => {
+    helper.unloadAgent(agent)
+    Object.keys(require.cache).forEach((key) => {
+      if (/pino/.test(key)) {
+        delete require.cache[key]
+      }
+    })
+  })
+
+  t.test('logging disabled', async (t) => {
+    setupAgent({ application_logging: { enabled: false } })
+    const disabledLogger = pino({ level: 'info' }, stream)
+    const message = 'logs are not enriched'
+    disabledLogger.info(message)
+    const line = await once(stream, 'data')
+    t.equal(line.msg, message, 'msg should not change')
+    t.equal(line.level, 30, 'level should not change')
+    t.match(line.time, /[0-9]{10}/, 'time should not change')
```

Comment:
yes because pino outputs the time as a key of `time`. if our application logging is enabled it renames this to timestamp. I could just change it to verify the key still exists